micro: port operator LEAKY_RELU kernel from lite with test

ddavis-2015 opened the pull request:
Complete implementation of TFLM operator LEAKY_RELU and associated TFLM test code.

PR step 5 of the work to port operator LEAKY_RELU as tracked in Issue #46161

ddavis-2015 committed:
micro: port operator LEAKY_RELU kernel from lite with test

Complete implementation of TFLM operator LEAKY_RELU and associated TFLM test code.

PR step 5 of the work to port operator LEAKY_RELU as tracked in Issue #46161

gbaned commented:
@ddavis-2015  This PR is in draft, any update on this? Please. Thanks!

advaitjain commented on tensorflow/lite/micro/kernels/leaky_relu.cc:
```diff
@@ -28,55 +28,59 @@ namespace micro {
+constexpr int kInputTensor = 0;
+constexpr int kOutputTensor = 0;
+
 // OLD-TODO(b/142762739): We should figure out a multi-threading plan for most
 // of the activation ops below.
```
remove, it is not relevant to TFLM.

ddavis-2015 replied:
Removed

advaitjain commented on tensorflow/lite/micro/kernels/leaky_relu.cc:
```diff
@@ -28,55 +28,59 @@ namespace micro {
+struct LeakyReluOpData {
+  // quantization parameters
+  int32_t output_multiplier_alpha;
+  int output_shift_alpha;
+  int32_t output_multiplier_identity;
+  int output_shift_identity;
```
I think we should keep these as int32_t. We do have some non-32 bit targets as well.

ddavis-2015 replied:
Fixed

advaitjain commented on tensorflow/lite/micro/kernels/leaky_relu.cc:
```diff
@@ -28,55 +28,59 @@ namespace micro {
 template <typename T>
-void QuantizeLeakyRelu(const TfLiteTensor* input, TfLiteTensor* output,
+void QuantizeLeakyRelu(const TfLiteEvalTensor* input, TfLiteEvalTensor* output,
                        const LeakyReluOpData* data) {
```
The TfLite code doesn't always follow the Google style guide and we are incrementally moving towards better conformance within TFLM.


Reorder the params to be inputs then outputs. And consider changing the params to be const and non-const references (this may be awkward given the existing APIs, but do it if possible).

Same suggestion for the other (non-kernel API) functions in this kernel and the test as well.

https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

ddavis-2015 replied:
Fixed

advaitjain commented on tensorflow/lite/micro/kernels/leaky_relu.cc:
```diff
@@ -92,52 +96,73 @@ TfLiteStatus LeakyReluPrepare(TfLiteContext* context, TfLiteNode* node) {
+    }
     case kTfLiteInt8: {
       QuantizeLeakyRelu<int8_t>(input, output, data);
       return kTfLiteOk;
-    } break;
+    }
     case kTfLiteInt16: {
       QuantizeLeakyRelu<int16_t>(input, output, data);
       return kTfLiteOk;
-    } break;
+    }
     default:
-      TF_LITE_KERNEL_LOG(
-          context,
-          "Only float32, int8, int16 and uint8 is supported currently, got %s.",
-          TfLiteTypeGetName(input->type));
+      TF_LITE_KERNEL_LOG(context,
```
only add support for Int8 and Float32 to begin with. Additional data-type support should be part of follow-on PR.

ddavis-2015 replied:
Fixed

advaitjain commented on tensorflow/lite/micro/kernels/leaky_relu.cc:
```diff
@@ -92,52 +96,73 @@ TfLiteStatus LeakyReluPrepare(TfLiteContext* context, TfLiteNode* node) {
     TF_LITE_ENSURE_EQ(context, output->params.zero_point, 0);
```
its somewhat awkward but there are some implicit float to double conversions happening that break the xtensa build (not part of the CI right now), and we will need some casting to get blupill to pass if we keep the structs as all int32_t.

Here's what I did after checking out this PR:
```cc
    double alpha_multiplier = static_cast<double>(input->params.scale) *
                              static_cast<double>(params->alpha) /
                              static_cast<double>(output->params.scale);

    int output_shift_alpha;
    QuantizeMultiplier(alpha_multiplier, &data->output_multiplier_alpha,
                       &output_shift_alpha);
    data->output_shift_alpha = static_cast<int32_t>(output_shift_alpha);

    double identity_multiplier = static_cast<double>(input->params.scale) /
                                 static_cast<double>(output->params.scale);
    int output_shift_identity;
    QuantizeMultiplier(identity_multiplier, &data->output_multiplier_identity,
                       &output_shift_identity);
    data->output_shift_identity = static_cast<int32_t>(output_shift_identity);
```

The static_cast<double> are only needed with the xtensa toolchain (since it is not reproducible for you, if you miss some of them, that is ok).

The static_cast<int32_t> is for the bluepill target.

I'm going to see if it is possible to get the double promotion errors on xtensa to also be reproducible with either arm-gcc or gcc / clang. If not, I can fix any remaining issues separate from this PR.

ddavis-2015 replied:
Fixed

advaitjain commented on tensorflow/lite/micro/kernels/leaky_relu.cc:
```diff
@@ -28,55 +28,59 @@ namespace micro {
 namespace activations {
```
we are going with a flat tflite namespace instead of the nested namespaces that are common in the TfLite code.

Some context: https://abseil.io/tips/130

ddavis-2015 replied:
Fixed

gbaned commented:
@ddavis-2015  Can you please check @advaitjain's comments and resolve conflicts?. Thanks!

ddavis-2015 committed:
Merge branch 'master' into LeakyRelu-pr5

ddavis-2015 committed:
fix review issues

Flatten namespace
Reorder input/output parameters
Explicit int/int32_t and float/double conversions

ddavis-2015 committed:
Support only float32/int8

petewarden commented on tensorflow/lite/micro/kernels/leaky_relu.cc:
```diff
@@ -23,122 +23,129 @@ limitations under the License.
+                               tflite::micro::GetTensorShape(output),
+                               tflite::micro::GetTensorData<float>(output));
       return kTfLiteOk;
-    } break;
```
A bit of a nit-pick, but could you add "break" back in here, and below? I know the return earlier makes it redundant, but if the body of the case statement was changed in the future, we'd end up with an accidental fall through.

ddavis-2015 replied:
Fixed.

petewarden reviewed (changes requested):
Added one comment, and can you also add this to all_ops_resolver.cc?

ddavis-2015 committed:
add LEAKY_RELU to AllOpsResolver

ddavis-2015 committed:
restore break statement to end of case block

ddavis-2015 committed:
Merge branch 'master' into LeakyRelu-pr5

ddavis-2015 committed:
Merge branch 'master' into LeakyRelu-pr5
